fix: `utils.UpdateObject` duplication of array members

matt-FFFFFF opened the pull request:
fixes #545

`utils.UpdateObject` would duplicate array members under certain conditions.

Adds `reflect.DeepEqual()` check to start of `utils.UpdateObject` to return early if old is the same as new. This could be masking the original bug but it's still probably a worthy thing to do.

This PR adds a test, which used to fail but now passes.

Pinging @ms-henglu

mcleanbc commented:
I also ran into the linked issue, using the Microsoft.CognitiveServices/accounts/raiPolicies API, which returns multiple entries with the same value for `name`, e.g.

```
               "contentFilters": [
                    {
                        "name": "Hate",
                        "allowedContentLevel": "Medium",
                        "blocking": true,
                        "enabled": true,
                        "source": "Prompt"
                    },
                    {
                        "name": "Hate",
                        "allowedContentLevel": "Medium",
                        "blocking": true,
                        "enabled": true,
                        "source": "Completion"
                    }, 
```

If the API returns the `contentFilters` array in guaranteed order, the return early solution would work: the array would be deeply equal. But if the array order is not guaranteed (for this API or others with similar treatment of `name`), the issue would reoccur.

Rather than (or in addition to) a return early, I'd recommend replacing the call to `areSameArrayItems` on [ln 105](https://github.com/matt-FFFFFF/terraform-provider-azapi/blob/66650c8495d270697379ef3abb4f02d4a15b43f7/utils/json.go#L105) with `reflect.DeepEqual` which would be more reliable than checking the `name` field, which is the current behavior. https://github.com/matt-FFFFFF/terraform-provider-azapi/blob/66650c8495d270697379ef3abb4f02d4a15b43f7/utils/json.go#L133

matt-FFFFFF commented:
Thanks for this @mcleanbc 

I'll update my PR with your suggestions and re-test.

ms-henglu commented:
Thanks @matt-FFFFFF for the PR! And thanks @mcleanbc for the great suggestions!

mcleanbc commented:
Looks good to me. Thanks @matt-FFFFFF for including those updates!

ms-henglu commented:
Hello @matt-FFFFFF , thanks for working on this.

But the test case should not be modified. Let me add more contexts here.

The logic uses `name` field as identifier is used to solve the below issue.

User uses azapi_resource to manage a aks cluster, which must have a system node pool defined. And he also uses azapi_resource to add a user node pool. 

```hcl
resource "azapi_resource" "cluster" {
  type      = "Microsoft.ContainerService/managedClusters@2024-04-02-preview"
  parent_id = azapi_resource.resourceGroup.id
  name      = "example"
  location  = azapi_resource.resourceGroup.location
  identity {
    type         = "SystemAssigned"
    identity_ids = []
  }
  body = {
    properties = {
      agentPoolProfiles = [
        {
          count  = 1
          mode   = "System"
          name   = "default"
          vmSize = "Standard_DS2_v2"
        },
      ]
      dnsPrefix = "example"
    }
  }
}

resource "azapi_resource" "pool" {
  type      = "Microsoft.ContainerService/managedClusters/agentPools@2024-04-02-preview"
  parent_id = azapi_resource.cluster.id
  name      = "internal"
  body = {
    properties = {
      count  = 1
      mode   = "User"
      vmSize = "Standard_DS2_v2"
    }
  }
}
```

In the next plan, the `azapi_resource.cluster` will say there's diff in the `body.properties.agentPoolProfiles`, because there're actually two elements in the remote, but only one is defined. 

```bash
 ~ agentPoolProfiles = [
                  ~ {
                      - currentOrchestratorVersion = "1.28.9"
                      - enableFIPS                 = false
                      - kubeletDiskType            = "OS"
                      - maxPods                    = 110
                        name                       = "default"
                      - nodeImageVersion           = "AKSUbuntu-2204gen2containerd-202406.19.0"
                      - orchestratorVersion        = "1.28"
                      - osDiskSizeGB               = 128
                      - osDiskType                 = "Managed"
                      - osSKU                      = "Ubuntu"
                      - osType                     = "Linux"
                      - powerState                 = {
                          - code = "Running"
                        }
                      - provisioningState          = "Succeeded"
                      - type                       = "VirtualMachineScaleSets"
                      - upgradeSettings            = {
                          - maxSurge = "10%"
                        }
                        # (3 unchanged attributes hidden)
                    },
                  - {
                      - count                      = 1
                      - currentOrchestratorVersion = "1.28.9"
                      - enableAutoScaling          = false
                      - enableFIPS                 = false
                      - kubeletDiskType            = "OS"
                      - maxPods                    = 110
                      - mode                       = "User"
                      - name                       = "internal"
                      - nodeImageVersion           = "AKSUbuntu-2204gen2containerd-202406.19.0"
                      - orchestratorVersion        = "1.28"
                      - osDiskSizeGB               = 128
                      - osDiskType                 = "Managed"
                      - osSKU                      = "Ubuntu"
                      - osType                     = "Linux"
                      - powerState                 = {
                          - code = "Running"
                        }
                      - provisioningState          = "Succeeded"
                      - scaleDownMode              = "Delete"
                      - type                       = "VirtualMachineScaleSets"
                      - vmSize                     = "Standard_DS2_v2"
                    },
                ]
```

The goal is to make it easier for users to use `lifecycle.ignore_changes` to suppress the diff. And we don't want users to use `lifecycle.ignore_changes=[body.properties.agentPoolProfiles]` to suppress the whole array which makes it impossible to update the system node pool.

So I use the `name` field as an identifier, to map the system node pool and make it use the same schema as the config. Then the plan-diff will only show there's only one element added, and the system node pool has no diff. Then user can use `lifecycle.ignore_changes=[body.properties.agentPoolProfiles[1]]` to suppress the diff.

```bash
agentPoolProfiles = [
                    {
                        count  = 1
                        mode   = "System"
                        name   = "default"
                        vmSize = "Standard_DS2_v2"
                    },
                  - {
                      - count                      = 1
                      - currentOrchestratorVersion = "1.28.9"
                      - enableAutoScaling          = false
                      - enableFIPS                 = false
                      - kubeletDiskType            = "OS"
                      - maxPods                    = 110
                      - mode                       = "User"
                      - name                       = "internal"
                      - nodeImageVersion           = "AKSUbuntu-2204gen2containerd-202406.19.0"
                      - orchestratorVersion        = "1.28"
                      - osDiskSizeGB               = 128
                      - osDiskType                 = "Managed"
                      - osSKU                      = "Ubuntu"
                      - osType                     = "Linux"
                      - powerState                 = {
                          - code = "Running"
                        }
                      - provisioningState          = "Succeeded"
                      - scaleDownMode              = "Delete"
                      - type                       = "VirtualMachineScaleSets"
                      - vmSize                     = "Standard_DS2_v2"
                    },
                ]
```

ms-henglu commented on utils/json.go:
```diff
@@ -102,7 +106,7 @@ func UpdateObject(old interface{}, new interface{}, option UpdateJsonOption) int
 
 			for _, oldItem := range oldValue {
```
How about the below codes:
It tries to find the deeply equaled element first, if not found, use the item with same identifier.

```go
for _, oldItem := range oldValue {
	found := false
	for index, newItem := range newArr {
		if reflect.DeepEqual(oldItem, newItem) && !used[index] {
			res = append(res, UpdateObject(oldItem, newItem, option))
			used[index] = true
			found = true
			break
		}
	}
	if found {
		continue
	}
	for index, newItem := range newArr {
		if areSameArrayItems(oldItem, newItem) && !used[index] {
			res = append(res, UpdateObject(oldItem, newItem, option))
			used[index] = true
			break
		}
	}
}
```

matt-FFFFFF commented:
Thanks @ms-henglu I have updated the code with your suggestion.

I have reverted the changes to the test case and also added new cases to cover the scenarios raised by myself and @mcleanbc.

All tests now pass.

ms-henglu reviewed (approved):
Thanks for the PR! LGTM

matt-FFFFFF commented:
Thanks for approving! please can you make a `v1.14.1` patch release if `v1.15.0` is not ready yet?
